learnEcl/2700-function.md

```diff
@@ -2,3 +2,147 @@
 title: Function
 slug: function
 ---
+
+# FUNCTION
+
+A Function is a set of statements that take inputs, does some specific computation and produces an output or return a result. Result could be a value or a dataset.
+
+Notes
+
+* For function to be called/used from outside, EXPORT is required
+* Function name should match the file name. if not "Error: Definition must contain EXPORT or SHARED " is generated
+
+## Syntax
+
+<pre>
+<EclCode code="EXPORT [return_data_type] function_name (data_type arg [, data_type arg]+) := FUNCTION
+
+    ecl_code
+    RETURN return_value;
+
+END;">
+</EclCode>
+</pre>
+
+| _Value_ | _Definition_ |
+| :- | :- |
+| EXPORT | Optional |
+| return_data_type | Optional (compiler can infer it from return_value). If returning a dataset, the data type is DATASET(record_definition) |
+| function_name | The name by which the function will be invoked |
+| data_type | The argument’s data type. If passing a dataset, the data type is DATASET(record_definition) |
+| ecl_code | Whatever code is needed to build return_value. Conversely, if the code does not contribute to return_value then it is ignored. Attributes defined here are scoped to the function |
+| RETURN | Required |
+| return_value | 	The result of the function |
+| END | Required |
+
+**Example**
+
+<pre>
+<EclCode
+id="FunctionExp_1"
+tryMe="FunctionExp_1"
+code="/*Function Example:*/
+
+/*
+Function Example:
+*/
+
+EXPORT MyFunc (STRING Val) := FUNCTION
+
+   Result := 'Hello ' + Val + ' , welcome to this function';
+   RETURN Result;
+
+END;
+
+// Using MyFunc
+Res := MyFunc('Jonny');
+OUTPUT(Res, NAMED('res'));
+
+OUTPUT(MyFunc('Sunny'), NAMED('Sunny'));
+
+"></EclCode>
+</pre>
+
+## Outputs in Function - Using WHEN
+
+OUTPUT can be used to return multiple results from a function. PARALLEL and WHEN are the keywords used to generate multiple results.
```

Comment:
"Return multiple results from a function" does not mean the same as "OUTPUT multiple values".  Values that are outputted are not available to the caller, for instance.

PARALLEL does not return multiple results at all.

WHEN is used to kick off an action after a value is computed.